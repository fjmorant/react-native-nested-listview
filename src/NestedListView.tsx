/* @flow */

import isEqual from 'lodash.isequal'
import * as React from 'react'
import {StyleSheet, Text, View} from 'react-native'
import * as shortid from 'shortid'
import NodeView, {INode} from './NodeView'

const styles = StyleSheet.create({
  errorContainer: {
    borderColor: 'rgb(84, 85, 86)',
    backgroundColor: 'rgb(237, 57, 40)',
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center',
    height: 60,
  },
  errorText: {
    color: 'rgb(255, 255, 255)',
    fontSize: 17,
    fontWeight: 'bold',
  },
})

export interface IProps {
  data: any
  extraData?: any
  renderNode: (elem: any, level?: number) => any
  onNodePressed?: () => any
  getChildrenName: (elem: any) => any
  style?: any
}

export interface IState {
  root: any
}

export default class NestedListView extends React.PureComponent<
  IProps,
  IState
> {
  public componentWillMount() {
    this.setState({root: this.generateRootNode(this.props)})
  }

  public componentWillReceiveProps(nextProps: any) {
    if (!isEqual(this.props.data, nextProps.data)) {
      this.setState({root: this.generateRootNode(nextProps)})
    }
  }

  public generateIds = (node?: INode) => {
    if (!node) {
      return
    }

    const childrenName: string = this.props.getChildrenName(node) || 'items'
    let children = node[childrenName]

    if (children) {
      if (!Array.isArray(children)) {
        children = Object.keys(children).map((key: string) => children[key])
      }
      node[childrenName] = children.map((_: INode, index: number) =>
        this.generateIds(children[index])
      )
    }

    node._internalId = shortid.generate()

    return node
  }

  public getChildrenName = (node: INode) => {
    if (node.name === 'root') {
      return 'items'
    }

    return this.props.getChildrenName
      ? this.props.getChildrenName(node)
      : 'items'
  }

  public renderErrorMessage(prop: string) {
    return (
      <View style={styles.errorContainer}>
        <Text style={styles.errorText}>prop {prop} has not been passed</Text>
      </View>
    )
  }

  public render() {
    const {data, getChildrenName, onNodePressed, renderNode} = this.props

    if (!getChildrenName) {
      return this.renderErrorMessage('getChildrenName')
    }

    if (!renderNode) {
      return this.renderErrorMessage('renderNode')
    }

    if (!data) {
      return this.renderErrorMessage('data')
    }

    return (
      <NodeView
        getChildrenName={this.getChildrenName}
        node={this.state.root}
        onNodePressed={onNodePressed}
        generateIds={this.generateIds}
        level={0}
        renderNode={renderNode}
        extraData={this.props.extraData}
      />
    )
  }

  private generateRootNode = (props: any): INode => {
    return {
      _internalId: shortid.generate(),
      items: props.data
        ? props.data.map((_: INode, index: number) =>
            this.generateIds(props.data[index])
          )
        : [],
      name: 'root',
      opened: true,
      hidden: true,
    }
  }
}
