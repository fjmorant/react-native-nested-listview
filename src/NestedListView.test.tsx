import * as React from 'react';
import { Text, View } from 'react-native';
import NestedListView, { NestedRow } from '.';
import { render, waitFor, fireEvent } from '@testing-library/react-native';

let mockCounter = 0;

jest.mock('shortid', () => ({
  generate: () => `${mockCounter++}`,
}));

const renderNode = (node: any) => (
  <View>
    <Text>{node.title}</Text>
  </View>
);

describe('NestedListView', () => {
  test('renders with simple array', async () => {
    const data = [
      { title: 'child1' },
      { title: 'child2' },
      { title: 'child3' },
    ];
    const { queryByText } = render(
      <NestedListView
        getChildrenName={() => 'items'}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title}</Text>
          </View>
        )}
        data={data}
      />,
    );

    await waitFor(() => {
      [1, 2, 3].forEach((item: number) => {
        const component = queryByText(`child${item}`);
        expect(component).toBeDefined();
      });
    });
  });

  test('renders with nested arrays', async () => {
    const data = [
      {
        title: 'child1',
        items: [{ name: 'subchild 1.1' }, { name: 'subchild 1.2' }],
      },
      { title: 'child2', items: [{ key: 'subchild 2.1' }] },
      { title: 'child3' },
    ];

    const { queryByText } = render(
      <NestedListView
        getChildrenName={() => 'items'}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title ?? node.name}</Text>
          </View>
        )}
        data={data}
      />,
    );

    await waitFor(() => {
      const parent = queryByText('child1');

      if (parent) {
        fireEvent.press(parent);
      }

      const component1 = queryByText('subchild 1.1');
      expect(component1).toBeDefined();

      const component2 = queryByText('subchild 1.2');
      expect(component2).toBeDefined();
    });
  });

  test('renders with nested arrays open and close', async () => {
    const data = [
      {
        title: 'child1',
        items: [{ name: 'subchild 1.1' }, { name: 'subchild 1.2' }],
      },
      { title: 'child2', items: [{ key: 'subchild 2.1' }] },
      { title: 'child3' },
    ];

    const { queryByText } = render(
      <NestedListView
        getChildrenName={() => 'items'}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title ?? node.name}</Text>
          </View>
        )}
        data={data}
      />,
    );

    await waitFor(() => {
      const parent = queryByText('child1');

      expect(queryByText('subchild 1.1')).toBeDefined();

      if (parent) {
        fireEvent.press(parent);
      }

      expect(queryByText('subchild 1.1')).toBeDefined();
    });
  });

  test('renders with nested arrays and children with different name', async () => {
    const data = [
      {
        title: 'child1',
        items: [{ name: 'subchild 1.1' }, { name: 'subchild 1.2' }],
      },
      { title: 'child2', descendants: [{ key: 'subchild 2.1' }] },
      { title: 'child3' },
    ];

    const { queryByText } = render(
      <NestedListView
        getChildrenName={(node: any) => {
          if (node.title === 'child2') {
            return 'descendants';
          }
          return 'items';
        }}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title}</Text>
          </View>
        )}
        data={data}
      />,
    );

    await waitFor(() => {
      const parent = queryByText('child2');

      if (parent) {
        fireEvent.press(parent);
      }

      const children = parent?.children;
      expect(children?.length).toEqual(1);
    });
  });

  test('renders with nested arrays and children with different name that does not exists', async () => {
    const data = [
      {
        title: 'child1',
        items: [{ name: 'subchild 1.1' }, { name: 'subchild 1.2' }],
      },
      {
        title: 'child2',
        descendants: [{ key: 'subchild 2.1', title: 'subchild 2.1' }],
      },
      { title: 'child3' },
    ];

    const { queryByText } = render(
      <NestedListView
        getChildrenName={(node: any) => {
          if (node.title === 'child2') {
            return 'children';
          }
          return 'items';
        }}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title}</Text>
          </View>
        )}
        data={data}
      />,
    );
    let parent: any;
    await waitFor(() => {
      parent = queryByText('child2');

      fireEvent.press(parent);

      const child = queryByText('subchild 2.1');

      expect(child).toBeNull();
    });
  });

  test('renders with children as objects', async () => {
    const data = [
      {
        name: 'Main Parent',
        children: {
          child1: {
            name: 'Main Child 1',
            children: {
              child1: {
                name: 'Sub Child 1',
                children: {},
              },
              child2: {
                name: 'Sub Child 2',
                children: {
                  subChild1: {
                    name: 'Sample',
                    children: {},
                  },
                },
              },
            },
          },
          child2: {
            name: 'Main Child 2',
            children: {
              child1: {
                name: 'Sub Child 1',
                children: {},
              },
              child2: {
                name: 'Sub Child 2',
                children: {},
              },
            },
          },
        },
      },
    ];

    const { queryByText } = render(
      <NestedListView
        getChildrenName={() => 'children'}
        renderNode={(node: any) => (
          <View>
            <Text>{node.name}</Text>
          </View>
        )}
        data={data}
      />,
    );

    await waitFor(() => {
      const parent = queryByText('Main Parent');

      if (parent) {
        fireEvent.press(parent);
      }

      const firstChild = queryByText('Main Child 1');
      expect(firstChild).toBeDefined();

      if (firstChild) {
        fireEvent.press(firstChild);
      }

      const secondChild = queryByText('Sub Child 2');
      expect(secondChild).toBeDefined();
    });
  });

  test('onNodePressed should be called when press a node', async () => {
    const data = [
      { title: 'child1' },
      { title: 'child2' },
      { title: 'child3' },
    ];

    const mockOnNodePressed = jest.fn();

    const { queryByText } = render(
      <NestedListView
        onNodePressed={mockOnNodePressed}
        renderNode={(node: any) => (
          <View>
            <Text>{node.title}</Text>
          </View>
        )}
        data={data}
      />,
    );

    const component = queryByText('child1');
    expect(component).toBeDefined();

    if (component) {
      fireEvent.press(component);
    }

    expect(mockOnNodePressed).toBeCalledTimes(1);
  });

  test('renders with NestedRow', async () => {
    const data = [
      { title: 'child1' },
      { title: 'child2' },
      { title: 'child3' },
    ];

    const mockOnNodePressed = jest.fn();

    const { UNSAFE_queryAllByType } = render(
      <NestedListView
        onNodePressed={mockOnNodePressed}
        renderNode={(node: any, level?: number) => (
          <NestedRow level={level}>{node.name}</NestedRow>
        )}
        data={data}
      />,
    );

    const component = UNSAFE_queryAllByType(NestedRow);
    expect(component.length).toEqual(3);
  });

  test('renders without renderNode', async () => {
    const data = [
      { title: 'child1' },
      { title: 'child2' },
      { title: 'child3' },
    ];

    const { queryByText } = render(
      <NestedListView
        // @ts-ignore
        renderNode={null}
        data={data}
      />,
    );

    const component = queryByText('prop renderNode has not been passed');
    expect(component).toBeDefined();
  });

  test('renders without data', async () => {
    const { queryByText } = render(
      <NestedListView
        renderNode={renderNode}
        // @ts-ignore
        data={null}
      />,
    );

    const component = queryByText('prop data has not been passed');
    expect(component).toBeDefined();
  });
});
