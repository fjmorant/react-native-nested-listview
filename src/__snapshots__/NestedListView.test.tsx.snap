// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedListView renders empty data 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
      },
      Object {
        "_internalId": "1",
      },
      Object {
        "_internalId": "1",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with NestedRow 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
              Object {
                "paddingLeft": 10,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
              Object {
                "paddingLeft": 10,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
              Object {
                "paddingLeft": 10,
              },
              Object {},
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with children as objects 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "children": Array [
          Object {
            "_internalId": "1",
            "children": Array [
              Object {
                "_internalId": "1",
                "children": Array [],
                "name": "Sub Child 1",
              },
              Object {
                "_internalId": "1",
                "children": Array [
                  Object {
                    "_internalId": "1",
                    "children": Array [],
                    "name": "Sample",
                  },
                ],
                "name": "Sub Child 2",
              },
            ],
            "name": "Main Child 1",
          },
          Object {
            "_internalId": "1",
            "children": Array [
              Object {
                "_internalId": "1",
                "children": Array [],
                "name": "Sub Child 1",
              },
              Object {
                "_internalId": "1",
                "children": Array [],
                "name": "Sub Child 2",
              },
            ],
            "name": "Main Child 2",
          },
        ],
        "name": "Main Parent",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text>
            Main Parent
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with nested arrays 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "items": Array [
          Object {
            "_internalId": "1",
            "name": "subchild 1.1",
          },
          Object {
            "_internalId": "1",
            "name": "subchild 1.2",
          },
        ],
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "items": Array [
          Object {
            "_internalId": "1",
            "key": "subchild 2.1",
          },
        ],
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with nested arrays and children with different name 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "items": Array [
          Object {
            "_internalId": "1",
            "name": "subchild 1.1",
          },
          Object {
            "_internalId": "1",
            "name": "subchild 1.2",
          },
        ],
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "descendants": Array [
          Object {
            "_internalId": "1",
            "key": "subchild 2.1",
          },
        ],
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with nested arrays and children with different that does not exists 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "items": Array [
          Object {
            "_internalId": "1",
            "name": "subchild 1.1",
          },
          Object {
            "_internalId": "1",
            "name": "subchild 1.2",
          },
        ],
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "descendants": Array [
          Object {
            "key": "subchild 2.1",
          },
        ],
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with onNodePressed 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders with simple array 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "_internalId": "1",
        "title": "child1",
      },
      Object {
        "_internalId": "1",
        "title": "child2",
      },
      Object {
        "_internalId": "1",
        "title": "child3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View>
          <Text />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`NestedListView renders without data 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(237, 57, 40)",
      "borderColor": "rgb(84, 85, 86)",
      "borderWidth": 1,
      "height": 60,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 17,
        "fontWeight": "bold",
      }
    }
  >
    prop 
    data
     has not been passed
  </Text>
</View>
`;

exports[`NestedListView renders without renderNode 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(237, 57, 40)",
      "borderColor": "rgb(84, 85, 86)",
      "borderWidth": 1,
      "height": 60,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 17,
        "fontWeight": "bold",
      }
    }
  >
    prop 
    renderNode
     has not been passed
  </Text>
</View>
`;
